arangod/Cluster/ClusterMethods.cpp

```diff
@@ -242,6 +242,103 @@ static void addTransactionHeaderForShard(transaction::Methods const& trx,
                                    "couldnt find shard in shardMap");
   }
 }
+
+/// @brief Collect the results from all shards (fastpath variant)
+///        All result bodies are stored in resultMap
+template <typename T>
+static void collectResponsesFromAllShards(
+    std::unordered_map<ShardID, std::vector<T>> const& shardMap,
+    std::vector<futures::Try<arangodb::network::Response>>& responses,
+    std::unordered_map<int, size_t>& errorCounter,
+    std::unordered_map<ShardID, std::shared_ptr<VPackBuilder>>& resultMap,
+    fuerte::StatusCode& code) {
+  // If none of the shards responds we return a SERVER_ERROR;
+  code = fuerte::StatusInternalError;
+  for (Try<arangodb::network::Response> const& tryRes : responses) {
+    network::Response const& res = tryRes.get();  // throws exceptions upwards
+    ShardID sId = res.destinationShard();
+
+    int commError = network::fuerteToArangoErrorCode(res);
+    if (commError != TRI_ERROR_NO_ERROR) {
+      auto tmpBuilder = std::make_shared<VPackBuilder>();
+      // If there was no answer whatsoever, we cannot rely on the shardId
+      // being present in the result struct:
+
+      auto weSend = shardMap.find(sId);
+      TRI_ASSERT(weSend != shardMap.end());  // We send sth there earlier.
+      size_t count = weSend->second.size();
+      for (size_t i = 0; i < count; ++i) {
+        tmpBuilder->openObject();
+        tmpBuilder->add(StaticStrings::Error, VPackValue(true));
+        tmpBuilder->add(StaticStrings::ErrorNum, VPackValue(commError));
+        tmpBuilder->close();
+      }
+      resultMap.emplace(sId, std::move(tmpBuilder));
+    } else {
+      std::vector<VPackSlice> const& slices = res.response->slices();
+      auto tmpBuilder = std::make_shared<VPackBuilder>();
+      if (!slices.empty()) {
+        tmpBuilder->add(slices[0]);
+      }
+
+      resultMap.emplace(sId, std::move(tmpBuilder));
+      network::errorCodesFromHeaders(res.response->header.meta, errorCounter, true);
+      code = res.response->statusCode();
+    }
+  }
+}
+
+OperationResult handleResponsesFromAllShards(
+    std::vector<futures::Try<arangodb::network::Response>>& responses,
+    std::function<void(Result&, VPackSlice)> handler) {
+  // If none of the shards responds we return a SERVER_ERROR;
+  Result result;
+  for (Try<arangodb::network::Response> const& tryRes : responses) {
+    network::Response const& res = tryRes.get();  // throws exceptions upwards
+    int commError = network::fuerteToArangoErrorCode(res);
+    if (commError != TRI_ERROR_NO_ERROR) {
+      result.reset(commError);
+      break;
+    } else {
+      std::vector<VPackSlice> const& slices = res.response->slices();
+      if (!slices.empty()) {
+        VPackSlice answer = slices[0];
+        handler(result, answer);
+        if (result.fail()) {
+          break;
+        }
+      }
+    }
+  }
+  return OperationResult(result);
+}
+
+template <typename T>
+std::pair<OperationResult, T> handleResponsesFromAllShards(
```

Comment:
I don't think you necessarily need this, the above version should be able to do the same thing ?

Reply:
This one allows you to return additional data instead of just the `OperationResult`. Otherwise you need to play some games with passing e.g. a dynamically allocated piece of data via `shared_ptr` capture. Basically this allows for more of a reduce-style handler. Does that make sense?

Reply:
In my view this would seem more natural
```suggestion
T val;
auto opres = handleResponsesFromAllShards(responses, 
[&val](VPackSlice slice) -> Result {
   val = doSomething(slice);
   return Result()
});
```